### PR TITLE
Grain activation local cache threading fix 

### PR DESCRIPTION
**Pull request by jason-bragg**
List of grain activations in local cache are now read only.  This forces immutable treatment of these lists, preventing enumeration and modification of the list from occurring at the same time.

The reported manifestation of the problem was that the call `GetLocalCacheData` was throwing an exception because it was iterating over the list of activations while that list was being modified in another thread.  This was fairly rare.

The `GetLocalCacheData` call is made in a grains turn while making a grain call.  This means it may be commonly called by many threads and needs to be thread safe.  It was not thread safe because the calls `AdjustLocalCache`, `RegisterAsync`, and `InvalidateCacheEntry` all modified the inner list with no thread synchronization.

Rather than adding thread synchronization, the list was marked read-only, so it could not be modified.  All code modifying the list was updated to respect the read-only constraint on the data.  When removing items from the list, instead of removing entries from the list in-place, we build a new list with the remaining items.  When adding items to the list, instead of adding them to the existing list, we make a new list containing the items from the original list, along with those we are adding.  Since the list can no longer be modified all readers are thread safe, because writers must replace the list, rather than modifying it.

There may still exist race conditions around the updating of the activation list, since changes are not protected, but this would only cause data consistency issues that the system will self correct.

It should also be noted that multiple activations of the same grain reference is supported by the grain directory but not currently supported by any Orleans grain types.  This means that while the activation lists support multiple activations, there really should only be a single activation in the list.  The code is written with some simple optimizations around this, which reduce the cost of maintaining readonly lists that must be duplicated.

### CODE REVIEW

**Comment by gabikliot** [comment]
Can you please provide more details on the original problem?
What lines exactly were creating the race? 

If there is a multi-threaded race, taking read-only copy may not help (or maybe yes, depending on the race), since the race might still occur when creating the copy. For me to be able to fully bless this fix, I need to understand more details of the original issue.

**Comment by jason-bragg** [comment]
@gabikliot 
>"provide more details"

Sure. :)  Description updated.

File: src/OrleansRuntime/GrainDirectory/AdaptiveGrainDirectoryCache.cs [hunk 1]
```diff
@@ -113,29 +113,26 @@ public void Clear()
...
+
+            NumHits++;      // for stats
+            tmp.NumAccesses++;
+            result = tmp.Value;
+            version = tmp.ETag;
+            return true;
         }
 
-        public List<Tuple<GrainId, TValue, int>> KeyValues
+        public IReadOnlyList<Tuple<GrainId, TValue, int>> KeyValues
```
Comment (gabikliot) [inline src/OrleansRuntime/GrainDirectory/AdaptiveGrainDirectoryCache.cs]: Just marking the return type as `IReadOnlyList` does  not make it readonly of course. It is just a marker interface.

**Comment by gabikliot** [comment]
Jason, thanks, I understand the original race now.
I do have a question about the solution.
You write:
" Rather than adding thread synchronization, the list was marked readonly, so it could not be modified."

Marking a list as read only does not make it read only of course (see here for more context: http://stackoverflow.com/questions/15262981/why-does-listt-implement-ireadonlylistt-in-net-4-5, but it is a pretty snandard thing in C#). The caller will not mutate, unless he casts it back, which he can. So in that regard just returning `IReadOnlyList` does not fix the problem compared to just returning the list as before (it is basically a no-op, except for marking intent). And also of course someone else, not the caller to this function, can still mutate the underlying list, while the caller is iterating. So this list is NOT read only.

Perhaps your other code does fix the problem (I am  not saying it does not). My comment here is about your explanation above. The explanation, by itself, does not explain the fix. Can you please provide a correct explanation for the fix, with more details? Since this is a rather complicated threading issue, I want us to be 100% on the same page w.r.t. to your fix.

**Comment by gabikliot** [comment]
Another comment about multiple vs. single activation: it is correct to say that most use cases for the directory are single activation, and thus it is OK, even good, to optimize. But just do keep in mind that we do have in the code now cases for multiple activations in the directory (client registers with multiple gateway activation ids). Thus it is important to maintain the correct implementation for multiple activations as well.

**Comment by sergeybykov** [comment]
@gabikliot 
>Just marking the return type as  IReadOnlyList  does not make it readonly of course.

Of course, the goal here is to better communicate the intent and to prevent accidental mutations of the lists.

>Perhaps your other code does fix the problem (I am not saying it does not).

Did you see where @jason-bragg added `new List<>` where it was previously being mutated in place?

https://github.com/dotnet/orleans/pull/1019/files#diff-580fa3b86dd0be4f1115aad7e2eb2d1bR346
https://github.com/dotnet/orleans/pull/1019/files?diff=unified#diff-580fa3b86dd0be4f1115aad7e2eb2d1bR592
https://github.com/dotnet/orleans/pull/1019/files?diff=unified#diff-580fa3b86dd0be4f1115aad7e2eb2d1bR818

**Comment by gabikliot** [comment]
Yes, I did see that. And that is why I wrote: "Perhaps your other code does fix the problem (I am not saying it does not)."
What I asked, **to help me code review it**, is an explanation of the fix. What Jason wrote now "Rather than adding thread synchronization, the list was marked readonly, so it could not be modified." is clearly not the right explanation. 

Again, I asked for the fix explanation:

> Can you please provide a correct explanation for the fix, with more details? Since this is a rather complicated threading issue, I want us to be 100% on the same page w.r.t. to your fix.

**Comment by gabikliot** [comment]
OK, the new explanation now make more sense. Sorry for the confusion. It is just very important to be detailed and clear in those explanations. I will take a look when I have time.

**Comment by gabikliot** [comment]
Yes, looks OK. 

I would probably suggest encapsulating the removal logic in a helper function, something like: `RemoveFromDirectoryList(DirectoryCache, IReadOnlyList<Tuple<SiloAddress, ActivationId>> list, Func<Tuple<SiloAddress, ActivationId>, bool> doRemove)`, to both reduce the code duplication (the 2 removal cases are quite similar) and also better encapsulate the intent of this special handling for removel (plus in the future the impl. of this `RemoveFromDirectoryList` may change, e.g., use lock).
But I am not insisting on this. Up to you to decide. Just to keep in mind.

**Comment by jdom** [comment]
Another option is to actually use an ImmutableList, where mutation is not even possible. https://msdn.microsoft.com/en-us/library/dn467185(v=vs.111).aspx
That requires adding a nuget dependency though, but it might be worth it in the long run, as it has some pretty nice immutable collections that might help in other cases in the future too.

**Comment by sergeybykov** [comment]
@jason-bragg Let's indeed eliminate the code duplication. If you don't feel like adding a dependency for `ImmutableList` here, let's rebase and merge this version.

**Comment by gabikliot** [comment]
By @veikkoeeva: 
https://gitter.im/dotnet/orleans?at=564c24aa3bff82362c336663
http://ayende.com/blog/164739/immutable-collections-performance

**Comment by jason-bragg** [comment]
>Immutable-collections

Initially looked into this, but I don't think this single usage justifies the dependency.  @gabikliot interesting info about performance.  Good to know for the future.

>Encapsulating the removal logic in a helper function

Started to structure it this way, but abandoned it for a couple reasons.  Will take another pass.

**Comment by jason-bragg** [comment]
Added utility function for removing activations from cache.
Rebased.
Squashed.
/dance

**Comment by gabikliot** [comment]
:+1:

**Pull request merged**